### PR TITLE
auto save session

### PR DESCRIPTION
**Pull request by cosminadrianpopescu**
Auto saves a session when loadFinished is emmited.

### CODE REVIEW

**Comment by The-Compiler** [comment]
I'm guessing the intention of this is to have some kind of crash recovery, right?

In any case, I think this is quite problematic - if we save the session before a crash, the user will just be in a crash loop (because usually, if a page crashes the browser, it'll do the same next time) and I'm pretty sure people will be frustrated by that (and I'll be by the support :laughing:) .

It'd be better to have some kind of special `_autosave` session, and then after a crash inform the user that they can load that session (after editing it if needed).

In the long run, ideally we'd have some kind of crash recovery dialog/page, where the user can select which tabs to restore.

**Comment by haasn** [comment]
A time-based autosave might be better. I've just bound save to `s` and have gotten into the habit of pressing it every now and then.

**Comment by cosminadrianpopescu** [comment]
Yes, the intention is to have a crash recovery. But the loop is not happening, because the session save is triggered on the `loadFinished` and if there is a crash, this happens before `loadFinished`, so the session is not saved. At least with the `webkit` engine. I've been using it for more than one year I think this feature, and never have a loop. 

However, for the loop not to happen, we could check that the session is not `default`, for example. Or maybe we could have a `run-safe` command line argument, which would start the `qutebrowser` with a clean new session.

**Comment by The-Compiler** [comment]
Oh, good point, I didn't consider that! I think `-R` (`--override-restore`) already is that commandline argument you're describing.

However, I'm still not sure if saving to the configured session is the best approach - for example, I'd rather control saving to non-default sessions by hand, but I'd still like to have crash recovery!

What about having an `_autosave` session which is periodically saved (probably without a configuration switch), and then deleted on a clean shutdown, and loaded on start when it's present?

**Comment by cosminadrianpopescu** [comment]
> What about having an _autosave session which is periodically saved (probably without a configuration switch), and then deleted on a clean shutdown, and loaded on start when it's present?

Yes, this would serve the same purpose. I will change the implementation to reflect this.

**Comment by The-Compiler** [comment]
Awesome, thanks!

File: qutebrowser/app.py [hunk 1]
```diff
@@ -210,14 +210,17 @@ def _load_session(name):
...
+    session_manager = objreg.get('session-manager')
     if name is None:
         try:
             name = state_config['general']['session']
         except KeyError:
             # No session given as argument and none in the session file ->
             # start without loading a session
-            return
-    session_manager = objreg.get('session-manager')
+            if session_manager.exists('_autosave'):
```
Comment (cosminadrianpopescu) [inline qutebrowser/app.py]: Here, I've been giving priority to the session saved by the user. This means, that if the user quits with `:wq`, then restarts `qutebrowser`, and then `qutebrowser` crashes for some reason, next time `qutebrowser` is restarted, the session saved lastly with `wq` will be saved, and not the one autosaved.

What do you think?

Comment (The-Compiler) [inline qutebrowser/app.py]: What's the rationale for that? I can't really see the point - if I do `qutebrowser -r foo` and use it for an hour (but never save my session explicitly) and it crashes, surely I'd want the autosaved sessions to be loaded, no?

File: qutebrowser/app.py [hunk 2]
```diff
@@ -723,6 +726,7 @@ def _shutdown(self, status, restart):
         # Now we can hopefully quit without segfaults
         log.destroy.debug("Deferring QApplication::exit...")
         objreg.get('signal-handler').deactivate()
+        objreg.get('session-manager').session_delete('_autosave', force = True)
```
Comment (The-Compiler) [inline qutebrowser/app.py]: Please don't uses spaces around `=` for function arguments (see [this section](https://www.python.org/dev/peps/pep-0008/#other-recommendations) of pep8)

File: qutebrowser/browser/browsertab.py [hunk 3]
```diff
@@ -685,6 +685,8 @@ def _auto_insert_mode_cb(elem):
 
     @pyqtSlot(bool)
     def _on_load_finished(self, ok):
+        sess_manager = objreg.get('session-manager')
+        sess_manager.session_save('_autosave', quiet = True, force = True)
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py]: Same as above :wink:

**Comment by cosminadrianpopescu** [comment]
> What's the rationale for that? I can't really see the point - if I do qutebrowser -r foo and use it for an hour (but never save my session explicitly) and it crashes, surely I'd want the autosaved sessions to be loaded, no?

You are right. I don't know what I was thinking. See the last commit.

File: qutebrowser/app.py [hunk 4]
```diff
@@ -723,6 +726,7 @@ def _shutdown(self, status, restart):
         # Now we can hopefully quit without segfaults
         log.destroy.debug("Deferring QApplication::exit...")
         objreg.get('signal-handler').deactivate()
+        objreg.get('session-manager').session_delete('_autosave', force=True)
```
Comment (The-Compiler) [inline qutebrowser/app.py]: `session_delete` is the command handler - this should call `.delete('_autosave')` instead (which doesn't need `force=True`), and make sure to handle `SessionError` and `OSError` (probably by just logging the exception).

File: qutebrowser/browser/browsertab.py [hunk 5]
```diff
@@ -685,6 +685,8 @@ def _auto_insert_mode_cb(elem):
 
     @pyqtSlot(bool)
     def _on_load_finished(self, ok):
+        sess_manager = objreg.get('session-manager')
+        sess_manager.session_save('_autosave', quiet=True, force=True)
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py]: Same as above, use `.delete('_autosave')`, not the command handler.

**Comment by The-Compiler** [comment]
Sorry for the delay - I just noticed you're using command handlers instead of calling the methods directly. After that's fixed, I think this should be ready.

**Comment by cosminadrianpopescu** [comment]
Sorry also for the delay. Holiday and other projects.

**Comment by The-Compiler** [comment]
Currently busy wth exams, but I'll get back to this in a few weeks.

**Comment by The-Compiler** [comment]
Thanks for the contribution and your patience! I added some error handling in ee25e900f8bba436f0ff18933b64655c593f8653 but otherwise this looks good.

**Pull request merged**